### ISSUE TITLE
Add Support for DebugRenderer

### ISSUE THREAD

**Issue by MAG-AdrianMeredith** [jrouwe/JoltPhysics.js#152]
Am I correct in thinking the debug renderer is not implemented, I can see the compile time flag but not reference to it in the IDL

**Comment by LeXXik** [comment]
Hi, @MAG-AdrianMeredith! Yep, correct. However, you can still implement debug renderer. You can check this answer in the Discussions:
https://github.com/jrouwe/JoltPhysics.js/discussions/49#discussioncomment-7799542

**Comment by MAG-AdrianMeredith** [comment]
Thanks man, that could work.  I've got a nasty issue with the collider diverging after an external transform is applied (we have gizmos moving/changing the shape) so i really need to see the actual collision world view of the model.  Tried to look at adding it to the idl but it doesn't look like a trivial thing for someone not familiar with it.  We also have a collision debug mode visualisation mode that need to be supported (like ammo)

May a request official support for the debug renderer?

**Comment by LeXXik** [comment]
That would be great. Exposing internal debug renderer would be beneficial, as the current method is a bit limited (e.g. can't see constraints).

**Comment by jrouwe** [comment]
The problem with the debug renderer is that having it costs performance even if you're not using it. WebIDL doesn't allow `#ifdef JPH_DEBUG_RENDERER` like C++ does, so I can't create a WebIDL and have a build with and without debug renderer. Also I'm not sure if I can properly express the DebugRenderer virtual interface in WebIDL.

Writing a loop that uses `Jolt.ShapeGetTriangles` in combination with `Jolt.BodyInterface.GetCenterOfMassTransform` is quite simple and should give you an exact representation of what is in the physics world.

**Comment by MAG-AdrianMeredith** [comment]
Thanks for the context, much appreciated.  However I can't see any reason the build.sh couldn't swap out the idl file at build time or even just macro it with search and replace wouldn't be too crazy imo.
The workaround **_should_** hopefully allow me to unblock my current issue but If I understand correctly it has a ton of other features that sound like they could be very useful (e.g. when things go flying for no reason) so theres still value in doing it.

p.s. thanks for maintaining this awesome lib

**Comment by PhoenixIllusion** [comment]
What I've done on another project is have the IDL concat into a temp IDL file in the CMake folder, and used that for the build process. That allows different build-config to use different IDL combinations, or a "minimal IDL" for smaller build config.

**Comment by PhoenixIllusion** [comment]
(more for people who want to look into this, not as anything prod-ready)
 
Implementing it myself (small use case), I see how adding all the detailed Render methods that are inside many of the individual already-existing classes could be difficult in any approach, especially if using WebIDL. I'm not sure if those are commonly used, or mainly just DrawBodies and DrawConstraints.

Enabling a simple DebugRenderSimple and just exposing the DrawBodies and DrawConstraints isn't too difficult to put inside it's own JPH_DEBUG_RENDER `#ifdef` but as DebugRenderSimple says in the comments, the render routine is slow for 1000's of lines vs using batches and buffering (stuttering on Falling Shapes). I probably need to implement a Cpp side caching and send all the lines to the JS side in large-chunk memory buffers.

```cpp
#ifdef JPH_DEBUG_RENDERER

#include "Jolt/Renderer/DebugRendererSimple.h"
		class JSDebugRenderer : public JPH::DebugRendererSimple
		{
		public:
			virtual void DrawLine(JPH::RVec3Arg inFrom, JPH::RVec3Arg inTo, JPH::ColorArg inColor) override
			{
				EM_ASM_INT({
						const renderer = Module['Renderer'];
						return renderer && renderer['DrawLine']($0, $1, $2, $3, $4, $5, $6);
				}, inFrom.mF32[0], inFrom.mF32[1], inFrom.mF32[2], inTo.mF32[0], inTo.mF32[1], inTo.mF32[2], inColor.mU32);
			}

			virtual void DrawTriangle(JPH::RVec3Arg inV1, JPH::RVec3Arg inV2, JPH::RVec3Arg inV3, JPH::ColorArg inColor, ECastShadow inCastShadow) override
			{
				EM_ASM_INT({
						const renderer = Module['Renderer'];
						return renderer && renderer['DrawTriangle']($0, $1, $2,   $3, $4, $5,   $6, $7, $8,  $9, $10);
				},
				inV1.mF32[0], inV1.mF32[1], inV1.mF32[2],
				inV2.mF32[0], inV2.mF32[1], inV2.mF32[2],
				inV3.mF32[0], inV3.mF32[1], inV3.mF32[2],
				inColor.mU32, inCastShadow);
			}

			virtual void DrawText3D(JPH::RVec3Arg inPosition, const string_view &inString, JPH::ColorArg inColor, float inHeight) override
			{
				EM_ASM_INT({
						const renderer = Module['Renderer'];
						return renderer && renderer['DrawText3D']($0, $1, $2,   $3, $4, $5);
				},
				inPosition.mF32[0], inPosition.mF32[1], inPosition.mF32[2],
				inString.data(),
				inColor.mU32, inHeight);
			}
		};
	extern "C" {
		EMSCRIPTEN_KEEPALIVE void debug_draw_bodies(PhysicsSystem* system) {
			system->DrawBodies(BodyManager::DrawSettings(), DebugRenderer::sInstance);
		}
		EMSCRIPTEN_KEEPALIVE void debug_draw_constraints(PhysicsSystem* system) {
			system->DrawConstraints(DebugRenderer::sInstance);
		}
	}
#endif
``` 

```typescript
interface DebugRoutines {
  _debug_draw_bodies(physicsSystemPtr: number): void;
  _debug_draw_constraints(physicsSystemPtr: number): void;
}
const Renderer = new DebugRender();
initJolt({Renderer}).then(function (Jolt) {
  // Initialize this example

  const debug = Jolt as any as DebugRoutines;
  initExample(Jolt, ()=>{
    debug._debug_draw_bodies(Jolt.getPointer(physicsSystem));
    debug._debug_draw_constraints(Jolt.getPointer(physicsSystem));
    Renderer.Finish(); // compile all vertex into buffers and push to cached or new meshes
  });
```
![2024-04-17-debug-render](https://github.com/jrouwe/JoltPhysics.js/assets/756488/8ff6cce2-aa43-4a22-b19d-685ebd7a2dc6)
![2024-04-17-debug-constraint](https://github.com/jrouwe/JoltPhysics.js/assets/756488/35ecddcd-e924-46a3-a4e9-431b0bbef340)

**Comment by LeXXik** [comment]
Personally, most of the time I need to debug draw is to resolve some collision issues, e.g. scale/orientation of the shape, etc. The vertices answer that for me, so everything else, like constraints, aabbs, axis, etc, land to "nice to have" closet. If I can see 2 bodies true positions (via vertex draw), I can figure out the constraint issues, even if I don't visually see it.

**Comment by jrouwe** [comment]
As an alternative: We could also always add the functions/classes to the IDL and on the CPP side just #ifdef the contents of the function out so that they become NOPs.

B.t.w. @PhoenixIllusion why did you use `EM_ASM_INT` construct instead of `JSImplementation` for the debug renderer?

We should probably move these drawing functions on the `JoltInterface` class for more consistency.

**Comment by PhoenixIllusion** [comment]
@jrouwe - I had not yet migrated the methods to WebIDL since I'd never used the Renderer before, and was still working with the code to try and make it compile, render, and disappear when targeting Distribution. I had started trying my approach of moving it to WebIDL during part of my lunch break.

I assume it would also need #ifdef the class constructor we could make it inherit off [nothing] and make sure no imports touch the Jolt Render .h headers since those trigger errors if the flag's undefined. I think as long as the compiler definition `JPH_DEBUG_RENDERER` is undefined, it should not cause any performance issues?

I was trying to keep an eye out for the use case:
> The problem with the debug renderer is that having it costs performance even if you're not using it.



I had some initial work setup copying the `DebugRendererRecorder` BatchImpl concept. The DebugRenderer would ask the JS implementation to convert vertex/triangle collections into uint32 tokens representing the Geometry ID. I was also going to add additional recording/batching of any line/triangle calls made during the execution of DrawBodies/DrawConstraints.

Is there a reason to prefer DrawBodies/DrawConstraints on JoltInterface? I was thinking of moving them to the DebugRenderer so anything drawing related was on that object.
I think it would go:

```typescript
initJolt().then(function (Jolt) {
  // Initialize this example
  const render = new Jolt.EmDebugRenderer();
  initExample(Jolt, ()=>{
    render.DrawBodies(physicsSystem);
    render.DrawConstraints(physicsSystem);
    render.Flush(); //unsure on best name here. we need to know a new frame should start next.
  });
```

**Comment by jrouwe** [comment]
> I assume it would also need #ifdef the class constructor we could make it inherit off [nothing] and make sure no imports touch the Jolt Render .h headers since those trigger errors if the flag's undefined. I think as long as the compiler definition JPH_DEBUG_RENDERER is undefined, it should not cause any performance issues?

Correct.

> Is there a reason to prefer DrawBodies/DrawConstraints on JoltInterface? I was thinking of moving them to the DebugRenderer so anything drawing related was on that object.

That's actually a better idea. The main thing is that I think they shouldn't be global functions (as we have none currently). (and now I see they were probably never global functions to begin with, the syntax just confused me)

**Comment by PhoenixIllusion** [comment]
IDL version Enabled on this branch https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/debug-renderer

I'm not sure on a good way to demo it, since it requires disabling the existing renderer if you want to see the meshes (or somehow messing with Z-buffer to force it visible.

The text-rendering works, but requires including the CSS3d extra from ThreeJS to see without some texture/canvas work. It only is required if the debug-render call uses DrawText3D calls (some constraints).

I'm not sure what some of the numbers are in the screenshots on Pulleys, or the giant white lines on Paths.

The debug-renderer.js doesn't have support for the Culling or DrawMode enum yet, though that will just require extending the material-cache so that it is some combination hash-key of `{color}-{drawmode}-{culling}` since those are all part of a material.

I think it also needs some cleanup code to handle entire colors worth of lines/triangles being deleted between frames, since all lines/triangles of a single color are combined into a single LineMesh and Triangle Mesh every frame.

```html
	<script src="js/three/CSS3DRenderer.js"></script>
	<script src="js/debug-renderer.js"></script>
```
The fact that the CSS rendering thing is a separate renderer makes it slightly awkward 
```typescript
const renderer = new DebugRenderer(Jolt);
initExample(Jolt, () => {
	lineUpdates.forEach(func => func());

	renderer.Initialize();
	renderer.DrawBodies(physicsSystem);
	renderer.DrawConstraints(physicsSystem);
	renderer.Finish();
	renderer.css3dRender && renderer.css3dRender.render(scene, camera);
});
```

![2024-04-19-debug-renderer-pulleys](https://github.com/jrouwe/JoltPhysics.js/assets/756488/1f7086e3-4682-4535-9314-56304aa2460a)
![2024-04-19-debug-renderer-svg-path](https://github.com/jrouwe/JoltPhysics.js/assets/756488/4154ca84-fc0b-4286-a063-183308ad2ab4)

**Comment by PhoenixIllusion** [comment]
I think a good demo would probably be wrapping up the options into a widgets with either toggles or checkboxes. The DebugRender can be on its own independent ThreeJS scene, and have a 3-way toggle between rendering Regular/Debug/Both scenes using the widget.

So a floating upper right widget, with a simple scene with a couple items and constraints. Probably some drop down for the color schemes.

The toggles would demonstrate things like "show velocities", "show supporting faces", "show contact manifolds", "toggle wireframe or solid". Im not sure about the soft body debugger, since it is a bit more complicated, and not quite synced with the JoltPhysics CPP yet for the color debugger on it.

Looking through on the CPP side outside of the demo, it seems a lot of the debug render options are static booleans spread over all the classes. Those can probably be wired to the DebugRender objects as Boolean getter/setter for easier access in JS.  Just searching the debug render define exposes most of them.

There are a couple other methods devoted debug render of skeletons and virtual characters (stair debug, stick to floor debug), and some RPM motor renderer.

**Comment by jrouwe** [comment]
> IDL version Enabled on this branch https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/debug-renderer

It's looking good!

Some remarks:

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/06641069-91d0-42c2-bf40-cedbb8968734)

These defines need to match up between the library and the glue code (if they don't you'll get crashes because you get function signature mismatches), with this change you're creating a mismatch. Jolt allows to change when the debug renderer is on through the DEBUG_RENDERER_IN_DEBUG_AND_RELEASE and DEBUG_RENDERER_IN_DISTRIBUTION cmake options.

In your current setup, the debug renderer is always off in the Distribution build (even though you can set the ENABLE_DEBUG_RENDERER cmake option). Maybe we should not expose the debug renderer as a separate option and just enable the DebugRenderer in Debug and Release builds as before (even though it was not exposed to the IDL before). The npm package uses the Distribution build so it won't have the DebugRenderer. That way the released package doesn't come with any overhead and if developers need to debug they can turn the DebugRenderer on.

> I'm not sure what some of the numbers are in the screenshots on Pulleys, or the giant white lines on Paths.

The number on the pulley is the current length of the rope. The white lines on the path constraints look like a bug (there's nothing that draws in white for the path constraint).

> Looking through on the CPP side outside of the demo, it seems a lot of the debug render options are static booleans spread over all the classes. Those can probably be wired to the DebugRender objects as Boolean getter/setter for easier access in JS. Just searching the debug render define exposes most of them.

I think most of these toggles are mainly interesting for me to debug the internals of the library and not so much for other people. E.g. the coloring of the soft bodies you're talking about shows me how the soft body has been divided up in batches that can be spread across threads, something that is totally uninteresting to 99% of the users of the library, same for the character virtual debug modes - they show internal details about how the interaction with the terrain is solved and without intimate knowledge of that code you wouldn't understand what they mean. For the ones that are interesting, like 'show velocities' I would indeed expose them on the DebugRenderer as properties.

**Comment by PhoenixIllusion** [comment]
I'd tried to hook into the existing `DEBUG_RENDERER_IN_DEBUG_AND_RELEASE` and `DEBUG_RENDERER_IN_DISTRIBUTION`, but given how verbose those CMake flags are, I will probably switch to your recommended "just tie it into the build targets, Debug/Release"

For a quick widget, I got something working, so it may be ready for a quick demo page and PR soon. For a demo with mobile support, it probably needs some support of expand/collapse.  The widget probably needs CSS polish, probably tinted background and padding/border

Some of the flags shown have dependencies internal to Jolt, specifically "Get Support Function + Support Direction" and "Shape + Shape Wireframe".

I used ThreeJS's layer support to set the scene to layer-mask-3 (0b011) and let the existing Example.js scene stay on layer 1, and debug on layer 2. That allows switching the camera to layer-masks  1, 2 or 3 for split/combining the renders. Combined looks horrible if you don't turn on wire framing, given everything z-fights all at once.

![2024-04-23-debug-render-checkboxes](https://github.com/jrouwe/JoltPhysics.js/assets/756488/48ccaa51-73bf-408e-a81c-38da1be15ca3)

**Comment by jrouwe** [comment]
Regarding the options: Get Support Function / Get Support Direction / Get Supporting Face are functions that even I barely use (they're created for me to be able to debug the functionality) and Predicted Bounds is also something you don't have to be concerned about as a user (Bounding Box should draw almost the exact same thing).

And a little typo: Intertia -> Inertia

Other than that it looks to me like it's ready to be integrated!

**Comment by jrouwe** [comment]
#157 was merged!
